.circleci/config.yml

```diff
@@ -1,63 +1,91 @@
-          name: Install the full development requirements
-          working_directory: ~/project/
-          command: pip install --progress-bar off -r requirements.txt
-      - run:
-          name: Show installed Python packages
-          command: pip list
+          name: Create custom requirements
+          command: grep << parameters.command-name >> requirements.txt > << parameters.command-name >>_requirements.txt
+      - python/install-packages:
+          pkg-manager: pip
+          pip-dependency-file: << parameters.command-name >>_requirements.txt
+          cache-version: << parameters.command-name >>-v1
       - run:
-          name: Lint
           working_directory: ~/
-          # Avoid being able to import wikipron by relative import.
-          # Test code by importing the *installed* wikipron in site-packages.
-          command: flake8 project/setup.py project/wikipron project/tests
+          command: << parameters.command-run >>
+
+  build-python:
+    parameters:
+      python-version:
+        type: string
+    docker:
+      - image: cimg/python:<< parameters.python-version >>
+        #auth:
+        #  username: $DOCKERHUB_USERNAME
+        #  password: $DOCKERHUB_PASSWORD
+    steps:
+      - checkout
       - run:
-          name: Check styling consistency with black
-          working_directory: ~/
-          # Avoid being able to import wikipron by relative import.
-          # Test code by importing the *installed* wikipron in site-packages.
-          command: black --line-length=79 --check project/setup.py project/wikipron project/tests project/data
+          name: Create empty pyenv file
+          command: |
+              mkdir -p /home/circleci/.pyenv/versions/
+              touch /home/circleci/.pyenv/version
+      - python/install-packages:
+          pkg-manager: pip
+          pre-install-steps:
+              - run:
+                  name: Build source distribution and install package from it
+                  command: |
+                      pip install --progress-bar off --upgrade pip setuptools
+                      python setup.py sdist
+                      pip install dist/*.tar.gz
+      - run:
+          name: Show installed Python packages
+          command: pip3 list -v
       - run:
           name: Run python tests
           working_directory: ~/
           # Avoid being able to import wikipron by relative import.
           # Test code by importing the *installed* wikipron in site-packages.
-          command: pytest -vv project/tests
-  build-python-3.7:
-    <<: *template
-    docker:
-      - image: python:3.7
-  build-python-3.8:
-    <<: *template
-    docker:
-      - image: python:3.8
-  build-python-3.9:
-    <<: *template
-    docker:
-      - image: python:3.9
+          command: |
+              sudo chown circleci:circleci /home
+              pytest -vv project/tests --junitxml /tmp/testxml/report.xml
+      - store_test_results:
+          path: /tmp/testxml/
```

Comment:
I notice that you've added `--junitxml /tmp/testxml/report.xml` to the pytest command above to generate a test report. Does this `store_test_results` step put the report somewhere? How do we access it?

Reply:
I added the `--junitxml /tmp/testxml/report.xml` and `store_test_results` because it is needed to split the test for the parallelization; you can access the results clicking on a build job and then selecting the test tab: [for example](https://app.circleci.com/pipelines/github/kylebgorman/wikipron/558/workflows/c28a52d8-0b7c-40bf-95a7-6ce66168d0d5/jobs/1630/tests)

Reply:
It it also useful to read failure messages cleanly formatted

Reply:
Interesting -- I must admit that I've been looking at test failures just by scrolling through the log on the "Steps" tab.